Make BO menu translatable

eternoendless opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Make the BO menu translatable through the translation interface and get rid of translation problems when installing other languages.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #9762 & Fixes #9893
| How to test?  | See below.

## About this change

This PR introduces several changes to the way automatic database content translation works.

### Understanding multi language entities, the menu, installer fixtures and automatic translation

Multi language entities in PrestaShop manage translated content by means of a second table, called `{entity_name}_lang`, where a copy of each one of the entity's multilingual fields are stored, translated to each of the installed languages.

Elements from the back office menu are stored in database and are managed by an entity called Tab. In order to have elements from the tab menu translated, the wordings from those "tabs" are multilingual.

On install, PrestaShop uses "data fixtures" (XML files containing predefined data) to populate certain tables, like shop configuration and elements of the BO menu, with default data. As a side note, this system is used for demo products as well.

During this process, wordings for multilingual entities are inserted on the database in the language chosen during install, or in English if translated fixtures for said data are not available (this is the case for Tab, which only has fixtures in English).

But as you have probably noticed, elements in the menu appear translated after you finish the translation in another language, like French. How does this happen?

PrestaShop has an _"automagic translation"_ feature for most multilingual entities. `Language::updateMultilangFromClass()` will loop through all tables finishing in `_lang`, and will try to translate their content using the appropriate class for that entity that extends `DataLang` (like `TabLang`, `ConfigurationLang`, `OrderStateLang`, and so forth) and is located in `classes/lang/`. 

Content is translated by performing a technique accurately described by @tomlev as _"untranslating and re-translating"_. It consists in taking the already translated content in the `lang` table, finding the original translation key by performing a reverse lookup in the translator component's dictionary, then using that key to translate the content into another language. 

This process is triggered during the install and whenever a new language is installed.

During install, Tab content is initially in English, regardless of your language of choice, because of fixtures. If you are installing in, say, French, the translator will contain a dictionary with pairs like `"Some Wording" => "Le wording en français"`. When the install process triggers this translation, it will find wordings like _"Some Wording"_ and will try to untranslate them. Since it won't find them, it will fall back to the original key. Then, it will perform a translation, resulting in _"Le wording en français"_, which is the value it will save.

If later on you install another language, like Spanish, then your dictionary will have pairs like `"Some Wording" => "El wording en español"`. Since content is saved in French, it will find values like _"Le wording en français"_ which results in _"Some Wording"_ when performing a reverse translation using the French dictionary. Using that key, we can translate the wording again using the Spanish dictionary, and we get _"El wording en español"_, which is saved in database.

### The problems

This sounds like a clever hack, and in most cases it works fine. However, in some cases two different wordings translate to the same phrase in a particular language. In those cases, a reverse translation could yield the wrong key, resulting in an erroneous translation to the target language.

The second problem appears in the specific case of the menu. Default menu entries are  supposed to be translatable via the translation interface in the Back Office (wordings are placed in the `Admin.Navigation.Menu` translation domain). However, translations modified through this interface are not reflected in the menu. This is because the automatic translation is not triggered after editing a translation, but even if it was, it would never work: the dictionary we're using to _"untranslate and retranslate"_ is the same we're updating. If the dictionary was updated before retranslation process, then the reverse lookup wouldn't work (because the menu name stored in database would be different from the one in the dictionary and therefore never found), and if the dictionary was updated after the retranslation process, then the previous value would be inserted in database, resulting in no change.

### The solutions

In my view, multilanguage data should be classified in two distinct groups. In the first one, wordings are stored as translation key and translation domain pairs, and are translated on the fly, exclusively via the translator component. In the second one, content is translated exclusively by the user via form (like, for example, a product's description). Mixing the two by performing untranslation and retranslation is an absurdity that results in too much complexity, which is at best unreliable.

This means that content which is not directly user-editable, like menu names, should always be translated using the translator, and never stored translated.

Ideally, this could be solved simply by adding a "wording" and "wording_domain" to each element in `tab`, and getting rid of `tab_lang` altogether. After that, it would be up to the presentation layer to get these properties and translate them on the fly using the translator. However, this would introduce a breaking change, since modules rely on the "name" property being stored translated. Therefore, this required a more complex solution.

In this change, I introduced the aforementioned columns, but kept the automatic translation feature so that the `name` column is still stored translated in `tab_lang` just as before. However, I added a specific behavior for Tab so that when the automatic translation is triggered, instead of performing a reverse translation lookup, it uses the provided wording and wording domain as basis for the retranslation. In order to provide zero-effort retrocompatibility, Tabs with no `wording` and `wording_domain` are translated using reverse lookup, just like before.

In addition, I refactored the retranslation feature a little so that I could trigger it on a single entity. This allows me to trigger a retranslation of Tabs whenever wordings from the menu domain (`Admin.Navigation.Menu`) are updated in the translation interface. 

Modules can also benefit from this system: they just need to fill out the `wording` and `wording_domain` fields on the tabs they create. If the wording exists in the catalogue, it will be translated. However, translating it from the back office interface won't trigger a retranslation unless it's attached to the `Admin.Navigation.Menu` domain. A future improvement could add a hook so that modules can add elements to the list of domains that trigger this.

**Updates on November 9, 2020:**

- Reworked how language packs are set up during the install 
  ⚠️ **WARNING:** This also includes that language packs are no longer downloaded again during install if already present in the `/translations` folder
- Removed a reference to "safe mode" during install, which no longer exists

**As noted on [April 8, 2021](https://github.com/PrestaShop/PrestaShop/pull/18082#issuecomment-815829907):**

Modules [need to be updated](https://devdocs.prestashop.com/1.7/modules/concepts/controllers/admin-controllers/tabs/#since-prestashop-178) to allow their tabs to be translated automatically when installing a new language.

## How to test

1. Install PrestaShop in French. Menu and other auto translatable things like order states should be in French.
2. Install PrestaShop in English, then add French. Menu and other auto translatable things like order states should be in French or English depending on your choice of language.
3. Edit a menu item in the BO translation page. After refreshing the page, the menu item should appear as translated.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/18082)
<!-- Reviewable:end -->

Commit:
Extract dynamic method names into variables

Commit:
Minor improvements

Commit:
Add wording and wording_domain to Tab entity

Commit:
Extract FileLoader from XmlLoader

Commit:
Add interface for all PS translators

Commit:
Introduce entity translators + add support for wording and wording domain in DataLang

Commit:
Minor improvements to FileLoader

Commit:
Make the menu translatable through the BO interface.

Trigger Tab entity translation when the domain used for menu wordings is modified though the translation interface.

Progi1984 commented:
@eternoendless Your PR put Travis in the :red_circle: zone !

Commit:
Add defaults for wording and wording_domain

Commit:
Use own prefix for test

Commit:
Allow null for wording and wording_domain

Commit:
Fix test

PierreRambaud commented on classes/Language.php:
```diff
@@ -1342,75 +1345,70 @@ public static function getLanguagePackListContent($iso, $tar)
     public static function updateMultilangTable($iso_code)
     {
-        $useLang = Db::getInstance()->getRow('SELECT * FROM `' . _DB_PREFIX_ . 'lang` WHERE `iso_code` = "' . pSQL($iso_code) . '" ', true, false);
+        $langId = Language::getIdByIso($iso_code);
```
```suggestion
        $langId = static::getIdByIso($iso_code);
```

PierreRambaud commented on classes/Language.php:
```diff
@@ -1342,75 +1345,70 @@ public static function getLanguagePackListContent($iso, $tar)
-        if (!empty($useLang)) {
-            $lang = new Language($useLang['id_lang']);
+        if (!empty($langId)) {
+            $lang = new Language($langId);
```
```suggestion
            $lang = new static($langId);
```

PierreRambaud commented on classes/Language.php:
```diff
@@ -1342,75 +1345,70 @@ public static function getLanguagePackListContent($iso, $tar)
+                ->getClassNameFromTable($tableName);
+
+            if (_DB_PREFIX_ . 'country_lang' === $tableName) {
+                self::updateMultilangFromCldr($language);
```
```suggestion
                static::updateMultilangFromCldr($language);
```

PierreRambaud commented on classes/Language.php:
```diff
@@ -1342,75 +1345,70 @@ public static function getLanguagePackListContent($iso, $tar)
+            if (_DB_PREFIX_ . 'country_lang' === $tableName) {
+                self::updateMultilangFromCldr($language);
+            } else {
+                self::updateMultilangFromClass($tableName, $className, $language);
```
```suggestion
                static::updateMultilangFromClass($tableName, $className, $language);
```

PierreRambaud commented on classes/Language.php:
```diff
@@ -1342,75 +1345,70 @@ public static function getLanguagePackListContent($iso, $tar)
+            }
+        }
+
+        Hook::exec('actionUpdateLangAfter', ['lang' => $language]);
```
Maybe add a 
`Hook::exec('actionUpdateLangBefore', ['lang' => $language]);` ?

eternoendless replied:
What for?

PierreRambaud commented on classes/Language.php:
```diff
@@ -1447,43 +1445,43 @@ public static function updateMultilangFromCldr($lang)
      * @param string $table
      * @param string $className
-     * @param string $lang
+     * @param Language $lang
```
```suggestion
     * @param static $lang
```
?

PierreRambaud commented on classes/Tab.php:
```diff
@@ -77,6 +83,8 @@ class TabCore extends ObjectModel
             'enabled' => ['type' => self::TYPE_BOOL, 'validate' => 'isBool'],
             'hide_host_mode' => ['type' => self::TYPE_BOOL, 'validate' => 'isBool'],
             'icon' => ['type' => self::TYPE_STRING, 'size' => 64],
+            'wording' => ['type' => self::TYPE_STRING, 'validate' => 'isString', 'size' => 194],
```
No need to put `194` if you don't add index on it :thinking:

PierreRambaud commented on classes/lang/TabLang.php:
```diff
@@ -23,12 +23,30 @@
+            $message = $value;
+        }
+
+        return $this->translator->trans($message, [], (!empty($domain)) ? $domain : $this->domain, $this->locale);
```
```suggestion
        return $this->translator->trans($message, [], (!empty($domain) ? $domain : $this->domain), $this->locale);
```
For a better readability

PierreRambaud commented on composer.json:
```diff
@@ -12,6 +12,7 @@
         "ext-json": "*",
         "ext-gd": "*",
         "ext-dom": "*",
+        "ext-simplexml": "*",
```
composer.lock needs to be updated

PierreRambaud commented on install-dev/data/xml/tab.xml:
```diff
@@ -5,450 +5,697 @@
         </tab>
 
         <!--SELL-->
         <tab id="SELL" id_parent="" active="1" enabled="1" hide_host_mode="0" icon="">
             <class_name>SELL</class_name>
```
Is these  `class_name` and `id` are ok? :thinking: 
```suggestion
        <tab id="Sell" id_parent="" active="1" enabled="1" hide_host_mode="0" icon="">
            <class_name>Sell</class_name>
```

eternoendless replied:
I don't remember what I found out about these but since it's unrelated I'd prefer to leave it as is for now. Be aware that this tab exists as it is since 1.7.0.

PierreRambaud commented on install-dev/data/xml/tab.xml:
```diff
@@ -5,450 +5,697 @@
+                        <wording>Stock Management</wording>
+                        <wording_domain>Admin.Navigation.Menu</wording_domain>
                     </tab>
                     <tab id="Stock_Movement" id_parent="Stock_Management" active="1" enabled="1" hide_host_mode="0" icon="">
```
```suggestion
                    <tab id="Stock_Movement" id_parent="Stock_Management_Parent" active="1" enabled="1" hide_host_mode="0" icon="">
```
?

PierreRambaud replied:
And for others

eternoendless replied:
Again, this is shadowy terrain that should definitely be explored, but not in this PR!

PierreRambaud commented on install-dev/data/xml/tab.xml:
```diff
@@ -5,450 +5,697 @@
 
 
         <!--IMPROVE-->
         <tab id="IMPROVE" id_parent="" active="1" enabled="1" hide_host_mode="0" icon="">
```
Again, an uppercase id :thinking:

PierreRambaud commented on install-dev/data/xml/tab.xml:
```diff
@@ -5,450 +5,697 @@
+                    <wording_domain>Admin.Navigation.Menu</wording_domain>
                 </tab>
                     <tab id="Mail_Theme" id_parent="Mail_Theme_Parent" active="1" enabled="1" hide_host_mode="0" icon="">
-                      <class_name>AdminMailTheme</class_name>
```
Indentation in the whole file is totally broken :sweat_smile:

eternoendless replied:
Yes because the structure makes no sense. This indentation is meant to reflect the menu structure which doesn't follow the xml structure (or is it the other way around? 🤪) 

Anyway I spent some hours trying to fix it and I found myself falling through the "refactor ALL the things!" rabbit hole, so I backtracked and decided to focus on the original intent.

PierreRambaud commented on install-dev/upgrade/sql/1.7.8.0.sql:
```diff
@@ -9,3 +9,6 @@ INSERT IGNORE INTO `PREFIX_hook` (`id_hook`, `name`, `title`, `description`, `po
   (NULL, 'displayBanner', 'Display Banner', 'Use this hook for banners on top of every pages', '1')
 ;
+
+ALTER TABLE `PREFIX_tab` ADD `wording` VARCHAR(196) DEFAULT NULL AFTER `icon`;
```
```suggestion
ALTER TABLE `PREFIX_tab` ADD `wording` VARCHAR(256) DEFAULT NULL AFTER `icon`;
```

PierreRambaud commented on install-dev/upgrade/sql/1.7.8.0.sql:
```diff
@@ -9,3 +9,6 @@ INSERT IGNORE INTO `PREFIX_hook` (`id_hook`, `name`, `title`, `description`, `po
 ;
+
+ALTER TABLE `PREFIX_tab` ADD `wording` VARCHAR(196) DEFAULT NULL AFTER `icon`;
+ALTER TABLE `PREFIX_tab` ADD `wording_domain` VARCHAR(196) DEFAULT NULL AFTER `wording`;
```
```suggestion
ALTER TABLE `PREFIX_tab` ADD `wording_domain` VARCHAR(256) DEFAULT NULL AFTER `wording`;
```

PierreRambaud commented on src/Adapter/EntityTranslation/DataLangClassNameNotFoundException.php:
```diff
@@ -0,0 +1,32 @@
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\EntityTranslation;
```
```suggestion
namespace PrestaShop\PrestaShop\Adapter\EntityTranslation\Exception;
```

Could it be in a separated directory?

PierreRambaud commented on src/Adapter/EntityTranslation/EntityTranslator.php:
```diff
@@ -0,0 +1,223 @@
+
+        $tableData = $this->db->executeS($sql, true, false);
+
+        if (!empty($tableData)) {
```
```suggestion
        if (empty($tableData)) {
            return;
        }
```

PierreRambaud commented on src/Adapter/EntityTranslation/EntityTranslator.php:
```diff
@@ -0,0 +1,223 @@
+                    $updateWhere = implode(' AND ', $updateWhere);
+                    $updateField = implode(', ', $updateField);
+
+                    $sql = "UPDATE `$tableNameSql` SET $updateField
```
Prepare the string outside the loop, and use sprintf instead. 
The `$shopFieldExists` shouldn't be executed on each loop.

eternoendless replied:
I'm refactoring the $shopFieldExists but leaving the string interpolation. AFAIK `sprintf` won't add anything and replacing named tokens with `%s` will decrease readability.

PierreRambaud commented on src/Adapter/EntityTranslation/EntityTranslator.php:
```diff
@@ -0,0 +1,223 @@
+     */
+    protected function getSourceString(array $data, string $fieldName): string
+    {
+        return $this->translator->getSourceString($data[$fieldName], $this->dataLang->getDomain());
```
Make sure `$data[$fieldName]` exists?

eternoendless replied:
Okay

PierreRambaud commented on src/PrestaShopBundle/Entity/Tab.php:
```diff
@@ -108,6 +108,20 @@ class Tab
+    /**
+     * @var string
+     *
+     * @ORM\Column(name="wording", type="string", length=194, nullable=true)
```
```suggestion
     * @ORM\Column(name="wording", type="string", length=256, nullable=true)
```

PierreRambaud commented on src/PrestaShopBundle/Entity/Tab.php:
```diff
@@ -108,6 +108,20 @@ class Tab
+    /**
+     * @var string
+     *
+     * @ORM\Column(name="wording_domain", type="string", length=194, nullable=true)
```
```suggestion
     * @ORM\Column(name="wording_domain", type="string", length=256, nullable=true)
```

PierreRambaud commented on src/PrestaShopBundle/Install/EntityLoader/FileLoader.php:
```diff
@@ -0,0 +1,143 @@
+            return self::FALLBACK_LANGUAGE_CODE;
+        }
+
+        return file_exists($this->langPath . $this->getFallbackToDefaultLanguage($iso) . '/data/' . $entity . '.xml') ? $iso :
```
You did thee times `file_exists` here. Why not checking only:
```suggestion
        return file_exists($this->langPath . $iso . '/data/' . $entity . '.xml') ? $iso : self::FALLBACK_LANGUAGE_CODE;
```

And you can remove `getFallbackToDefaultLanguage` method, and the previous check :thinking:

eternoendless replied:
Okay, but if for any reason this stops working you'll fix it 😛

marionf commented:
@eternoendless Does this PR will fix also #9816 and #9893 ?

eternoendless commented:
@marionf I think it should. However the "Link widget" may need an update in the module

Commit:
Use static references

Commit:
Use 255 for field size since they are not indexed

Commit:
Improve readability

Commit:
Move exception into exception namespace

Commit:
Use early return

Commit:
Refactor shop condition

Commit:
Make sure that the translated field exists in database record

Commit:
Reduce number of file_exists checks

Commit:
CS fix

eternoendless commented:
@PierreRambaud rebased and changes applied.

PierreRambaud commented on src/Adapter/EntityTranslation/EntityTranslator.php:
```diff
@@ -0,0 +1,238 @@
+    private function buildTableNameFromDataLang(DataLangCore $dataLang): string
+    {
+        $tableName = Inflector::tableize(get_class($dataLang));
+        if (substr($tableName, 0, strlen(_DB_PREFIX_)) !== _DB_PREFIX_) {
```
Use `construct` parameter for `dbprefix` like in `DataLangFactory`

eternoendless replied:
Done!

Commit:
Fix method signature

Commit:
Extract dbPrefix as a constructor parameter in EntityTranslator

Progi1984 commented on install-dev/data/xml/tab.xml:
```diff
@@ -5,450 +5,697 @@
                     </tab>
                 <tab id="Module_Page_old" id_parent="Modules" active="0" enabled="1" hide_host_mode="0" icon="">
                     <class_name>AdminModules</class_name>
+                    <wording/>
+                    <wording_domain/>
```
Normal ?

eternoendless replied:
Yes, this tab was never translated. I don't really know why it's here, notice it's `active="0"`.

Progi1984 replied:
Ok for me ;)

Progi1984 commented on src/Adapter/EntityTranslation/DataLangFactory.php:
```diff
@@ -0,0 +1,142 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
```suggestion
declare(strict_types=1);
```

Progi1984 commented on src/Adapter/EntityTranslation/EntityTranslator.php:
```diff
@@ -0,0 +1,246 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
```suggestion
declare(strict_types=1);

```

Progi1984 commented on src/Adapter/EntityTranslation/EntityTranslatorFactory.php:
```diff
@@ -0,0 +1,107 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
```suggestion
declare(strict_types=1);

```

Progi1984 commented on src/Adapter/EntityTranslation/Exception/DataLangClassNameNotFoundException.php:
```diff
@@ -0,0 +1,32 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
```suggestion
declare(strict_types=1);

```

Progi1984 commented on src/Adapter/EntityTranslation/TabTranslator.php:
```diff
@@ -0,0 +1,84 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
```suggestion
declare(strict_types=1);

```

Progi1984 commented on src/Core/Translation/EntityTranslatorInterface.php:
```diff
@@ -0,0 +1,41 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
```suggestion
declare(strict_types=1);

```

Progi1984 commented on src/PrestaShopBundle/Install/EntityLoader/FileLoader.php:
```diff
@@ -0,0 +1,164 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
```suggestion
declare(strict_types=1);

```

Progi1984 commented on src/PrestaShopBundle/Translation/TranslatorInterface.php:
```diff
@@ -0,0 +1,59 @@
+ */
+
+namespace PrestaShopBundle\Translation;
+
```
```suggestion

use Symfony\Component\Translation\TranslatorInterface as SymfonyTranslatorInterface;
```

Progi1984 commented on src/PrestaShopBundle/Translation/TranslatorInterface.php:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Interface for PrestaShop translators
+ */
+interface TranslatorInterface extends \Symfony\Component\Translation\TranslatorInterface
```
```suggestion
interface TranslatorInterface extends SymfonyTranslatorInterface
```

Progi1984 commented on src/PrestaShopBundle/Translation/TranslatorInterface.php:
```diff
@@ -0,0 +1,59 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
```suggestion
declare(strict_types=1);

```

Progi1984 commented on tests/Unit/Adapter/Translations/DataLangFactoryTest.php:
```diff
@@ -0,0 +1,59 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
```suggestion
declare(strict_types=1);

```

Progi1984 reviewed (commented):
Some changes

Commit:
Apply suggestions from code review

Add use statement

Co-authored-by: Progi1984 <progi1984@gmail.com>

Commit:
Apply suggestions from code review

Add missing strict types

Co-authored-by: Progi1984 <progi1984@gmail.com>

Progi1984 commented on src/PrestaShopBundle/Translation/TranslatorInterface.php:
```diff
@@ -0,0 +1,61 @@
+
+namespace PrestaShopBundle\Translation;
+
+use Symfony\Component\Translation\TranslatorInterface as SymfonyTranslatorInterface;
```
```suggestion
use Symfony\Component\Translation\TranslatorInterface as SymfonyTranslatorInterface;

```

Progi1984 committed:
Fixed Coding Style error

Progi1984 commented:
@eternoendless Some PHPCS fix errors ;) and Travis is now :green_circle:

SD1982 commented:
> ## How to test
> 1. Install PrestaShop in French. Menu and other auto translatable things like order states should be in French.
> 2. Install PrestaShop in English, then add French. Menu and other auto translatable things like order states should be in French or English depending on your choice of language.
> 3. Edit a menu item in the BO translation page. After refreshing the page, the menu item should appear as translated.

Hi @eternoendless If I reproduce step 2 everything is ok but if I do it in reverse i mean install in French => add English it doesn't seem to work properly the menu stays in french
https://drive.google.com/file/d/1U49TDB1kaQChUUp-7wR9p1DA-0kLeGy4/view?usp=sharing

I also have an exception when trying to update a langage : 
https://drive.google.com/file/d/1AO7eV7S_dqvPEZFsSEp0_JlRDXMXnfZl/view?usp=sharing

Pierstoval commented on src/PrestaShopBundle/Translation/DataCollectorTranslator.php:
```diff
@@ -29,7 +29,23 @@
+     */
+    public function clearLanguage($locale)
+    {
+        return  $this->__call('clearLanguage', [$locale]);
```
I'm not sure about these two methods. They seem to use the [base translator's __call() method](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Translation/DataCollectorTranslator.php#L113-L116) but it needs a decorated translator.

Why don't you call `$this->translator->clearLanguage($locale)`, since the decorated translator should anyway be an instance of Prestashop's one?
You could even type-hint this in the class with `/** @​property PrestashopTranslator $translator */` (of course you would need to import the namespace for auto-completion in IDEs).

eternoendless replied:
I had forgotten about this, the answer is pretty simple: the translator property is private on the original DataCollectorTranslator, so I can't access it 😢 

That said I think I could have used `parent::isLanguageLoaded()` and it would have been routed automagically all the same. But I don't know if it would have been clearer that way either anyway 😁

Pierstoval replied:
Calling `parent::isLanguageLoaded()` is definitely the way to go. Relying on magic methods is almost never a good option. Especially when the base method is reachable through `parent::` 🙂  (and especially in this case where it's more a hack than a true feature/pattern usage)

eternoendless replied:
I had to revert to this because PHPStan doesn't support calls to `parent` if the method is not declared  in the parent and routed with `__call()` ☹️ 

See https://github.com/PrestaShop/PrestaShop/pull/18082/checks?check_run_id=1375456805

Pierstoval replied:
@eternoendless PHPStan is not a true god that you have to follow! If you're certain that you have the right class at the right moment, and you tested the feature, then it works, and you can add an exception to PHPStan. If you do so, remember to add a small comment to explain why there's an exception :)

Commit:
Avoid reading all_languages.json from filesystem again and again

Commit:
Extract method & constants

Commit:
Replace self with static

Commit:
Remove safe mode block (removed in 1.7 - see Tools::getSafeModeStatus)

Commit:
Use early return on error

Commit:
Avoid redownloading the language pack on install

Commit:
Use separate method for first language pack install

During this phase the database is empty, so there's no use wasting cpu time updating it

Commit:
Fix missing wording for AdminTabs in install fixtures

Commit:
Fix crash when translating a DataLang whose FCQN does not match its table name

Commit:
Make magic method routing a little more clear

Commit:
Code style fixes

Commit:
Fix inverted argument order

eternoendless commented:
Rebased and bugs fixed.

- Reworked how language packs are set up during the install 
  ⚠️ **WARNING:** This also includes that language packs are no longer downloaded again during install if already present in the `/translations` folder
- Removed a reference to "safe mode" during install, which no longer exists
- Shuffled code around in Language class

### It works on my computer!

- Tested install in French, adding English, switching to that language, changing menu translations, installing more languages, uninstalling languages and reinstalling them again...
- Also verified that mail templates are in the correct language, as well as currencies
- Note: I installed English through localization pack, for all other languages I used _International > Translations > Add / update a language_

Progi1984 commented on classes/Language.php:
```diff
@@ -39,12 +42,23 @@ class LanguageCore extends ObjectModel implements LanguageInterface
     const ALL_LANGUAGES_FILE = '/app/Resources/all_languages.json';
     const SF_LANGUAGE_PACK_URL = 'https://i18n.prestashop.com/translations/%version%/%locale%/%locale%.zip';
     const EMAILS_LANGUAGE_PACK_URL = 'https://i18n.prestashop.com/mails/%version%/%locale%/%locale%.zip';
+    const PACK_TYPE_EMAILS = 'emails';
+    const PACK_TYPE_SYMFONY = 'sf';
```
```suggestion
    public const PACK_TYPE_EMAILS = 'emails';
    public const PACK_TYPE_SYMFONY = 'sf';
```

Commit:
Revert "Make magic method routing a little more clear"

This reverts commit eaf30ec391c433640a25fad0d46a0f4f5affca5e.

This kind of magic method call is not supported by PHPStan

Progi1984 commented on classes/Language.php:
```diff
@@ -1256,31 +1335,87 @@ public static function installEmailsLanguagePack($lang_pack, &$errors = [])
+     * @throws PrestaShopDatabaseException
+     * @throws PrestaShopException
+     */
+    public static function installFirstLanguagePack(string $iso, array $params = [], &$errors = []): bool
```
```suggestion
    public static function installFirstLanguagePack(string $iso, array $params = [], array &$errors = []): bool
```

Progi1984 commented on classes/lang/DataLang.php:
```diff
@@ -36,26 +38,33 @@
+     * @param string $locale
+     * @param TranslatorInterface|null $translator If defined, use this translator
+     */
+    public function __construct($locale, $translator = null)
```
```suggestion
    public function __construct($locale, ?TranslatorInterface $translator = null)
```

Commit:
Apply suggestions from code review

Co-authored-by: Progi1984 <progi1984@gmail.com>

Progi1984 commented on classes/Tab.php:
```diff
@@ -80,6 +86,8 @@ class TabCore extends ObjectModel
             'enabled' => ['type' => self::TYPE_BOOL, 'validate' => 'isBool'],
             'hide_host_mode' => ['type' => self::TYPE_BOOL, 'validate' => 'isBool'],
             'icon' => ['type' => self::TYPE_STRING, 'size' => 64],
+            'wording' => ['type' => self::TYPE_STRING, 'validate' => 'isString', 'size' => 255],
+            'wording_domain' => ['type' => self::TYPE_STRING, 'validate' => 'isString', 'size' => 255],
```
```suggestion
            'wording' => ['type' => self::TYPE_STRING, 'validate' => 'isString', 'allow_null' => true, 'size' => 255],
            'wording_domain' => ['type' => self::TYPE_STRING, 'validate' => 'isString', 'allow_null' => true, 'size' => 255],
```

Progi1984 commented on classes/Tab.php:
```diff
@@ -58,6 +58,12 @@ class TabCore extends ObjectModel
     /** @var string Icon font */
     public $icon;
 
+    /** @var string Wording to use for the display name */
+    public $wording;
+
+    /** @var string Wording domain to use for the display name */
+    public $wording_domain;
```
```suggestion
    /** @var string|null Wording to use for the display name */
    public $wording;

    /** @var string|null Wording domain to use for the display name */
    public $wording_domain;
```

Progi1984 commented on src/Adapter/EntityTranslation/EntityTranslator.php:
```diff
@@ -0,0 +1,246 @@
+        TranslatorInterface $translator,
+        DataLangCore $dataLang
+    ) {
+        //$this->translatorLanguageLoader = new TranslatorLanguageLoader(true);
```
```suggestion
```

Progi1984 commented on src/PrestaShopBundle/Entity/Tab.php:
```diff
@@ -108,6 +108,20 @@ class Tab
     private $icon;
 
+    /**
+     * @var string
```
```suggestion
     * @var string|null
```

Progi1984 commented on src/PrestaShopBundle/Entity/Tab.php:
```diff
@@ -108,6 +108,20 @@ class Tab
+    private $wording;
+
+    /**
+     * @var string
```
```suggestion
     * @var string|null
```

Progi1984 commented on src/PrestaShopBundle/Entity/Tab.php:
```diff
@@ -158,6 +172,22 @@ public function getTabLangs()
+    /**
+     * @return string|null
+     */
+    public function getWording()
```
```suggestion
    public function getWording(): ?string
```

Progi1984 commented on src/PrestaShopBundle/Entity/Tab.php:
```diff
@@ -158,6 +172,22 @@ public function getTabLangs()
+    /**
+     * @return string|null
+     */
+    public function getWordingDomain()
```
```suggestion
    public function getWordingDomain(): ?string
```

Julievrz commented:
Hi @eternoendless, the wordling looks good to me but I only spotted 2 strings. Is there something else that I might be missing? Thanks

eternoendless commented:
Don't worry @Julievrz I didn't change any of the existing wordings (nor did I add any new one).

eternoendless commented:
Rebased again

PierreRambaud commented on src/Adapter/EntityTranslation/EntityTranslator.php:
```diff
@@ -0,0 +1,245 @@
+                $updateWhere = implode(' AND ', $updateWhere);
+                $updateFields = implode(', ', $updateFields);
+
+                $sql = "UPDATE `$tableNameSql`
+                    SET $updateFields
+                    WHERE $updateWhere AND `id_lang` = $languageId"
+                    . $shopWhere
+                    . ' LIMIT 1';
```
```suggestion
                $sql = 'UPDATE `' . $tableNameSql . '` '
                    'SET $updateFields ' .
                    'WHERE ' . $updateWhere . ' AND `id_lang` = ' . $languageId . ' ' .
                    . $shopWhere
                    . ' LIMIT 1';
```

eternoendless replied:
There is an error in your suggestion. IMO I find the original more readable.

PierreRambaud replied:
I just don't like mixing ways to concatenate variables in strings :roll_eyes:

PierreRambaud commented on src/Adapter/EntityTranslation/EntityTranslator.php:
```diff
@@ -0,0 +1,245 @@
+    protected function shopFieldExists(string $tableNameSql): bool
+    {
+        $columns = $this->db->executeS(
+            sprintf('SHOW COLUMNS FROM `%s`', $tableNameSql)
```
```suggestion
            sprintf('SHOW COLUMNS FROM `%s`', bqSQL($tableNameSql))
```

PierreRambaud commented on src/Adapter/EntityTranslation/EntityTranslator.php:
```diff
@@ -0,0 +1,245 @@
+        );
+
+        foreach ($columns as $column) {
+            if ($column['Field'] == 'id_shop') {
```
```suggestion
            if ($column['Field'] === 'id_shop') {
```

PierreRambaud commented on src/PrestaShopBundle/Install/EntityLoader/FileLoader.php:
```diff
@@ -0,0 +1,165 @@
+     */
+    private function getLookupPaths(string $entity, ?string $iso = null): array
+    {
+        $fileName = "$entity.xml";
```
```suggestion
        $fileName = $entity . '.xml';
```

eternoendless replied:
I don't get this obsession with single quotes. I find interpolation much more readable.

PierreRambaud replied:
Just a habit, make sure the code is always written in the same way

Commit:
Apply suggestions from code review

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

Progi1984 commented:
@eternoendless Please rebase (PHPStan is :red_circle: & composer has conflicts).

eternoendless commented:
Rebased once again

Commit:
Fix call to undefined loadEntity method

Commit:
Code style fix

Commit:
Fix "Negated boolean expression is always false"

sowbiba commented on src/PrestaShopBundle/Translation/TranslatorInterface.php:
```diff
@@ -0,0 +1,62 @@
+{
+    /**
+     * Performs a reverse search in the catalogue and returns the translation key if found.
+     * AVOID USING THIS, IT PROVIDES APPROXIMATE RESULTS.
```
???

sowbiba reviewed (dismissed):
Rebase needed

eternoendless commented:
I'm not rebasing this again until it's been validated by QA

khouloudbelguith commented:
Hi @eternoendless,

This issue https://github.com/PrestaShop/PrestaShop/issues/9893 is not fixed with this PR
I cannot translate the `Stock` menu because it doesn't exist 

This issue: https://github.com/PrestaShop/PrestaShop/issues/9816 is not fixed with this PR
![image](https://user-images.githubusercontent.com/16067358/111203364-363dbf80-85c5-11eb-8052-70345ae79d8b.png)

But I can change the translation successfully 
![image](https://user-images.githubusercontent.com/16067358/111203586-7dc44b80-85c5-11eb-8c8f-579272a2f769.png)

Also, I found an error

1. Install shop with French language & France country
2. Go to BO > International > Localization page
3. Upload United states pack
4. Change employee language to English
5. The menu Design > Themes Catalog is still displayed in French Catalogue de thèmes
6. I tried to translate it but I found 
![image](https://user-images.githubusercontent.com/16067358/111203731-a5b3af00-85c5-11eb-8f02-233ee854d091.png)

I tried to change the translation of this expression in French, I found two similar expressions of `Catalogue de thèmes`
![image](https://user-images.githubusercontent.com/16067358/111204005-f88d6680-85c5-11eb-8372-d2e8bddd1c05.png)

Thank you!

Commit:
Fix Stock wording

Commit:
Allow wording and wording_domain in module tab register

eternoendless commented:
I Rebased the PR and added two commits:
- 15ede64 - which fixes the Stock wording
- 6c594e5 - which adds support for registering wording and wording domain in module tabs

@khouloudbelguith please test my PR again.

> This issue #9893 is not fixed with this PR
> I cannot translate the `Stock` menu because it doesn't exist

* The `Stock` wording is now fixed in this PR. 
* The other problem is fixed by in the Link Widget module: https://github.com/PrestaShop/ps_linklist/pull/116 (which requires this PR to work).

Regarding the "Theme Catalog" not being translated - this is actually due to the same problem that I fixed in Link Widget, but in this case it must be fixed in MBO. In fact, every module that adds a menu item will display the same problem and need to be fixed individually (ie. the module must register the Tab adding `wording` and `wording_domain`). The problem itself is unrelated to the PR, but this PR makes it possible to fix it!

Progi1984 commented on composer.lock:
```diff
@@ -58,6 +58,9 @@
                 "doctrine",
                 "orm"
             ],
+            "support": {
```
Please remove modifications on composer.lock

eternoendless replied:
These are added automatically by composer when you run `composer update nothing`
If I change it then composer.lock will become out of sync with composer.json

Progi1984 replied:
composer.lock is again out of sync

eternoendless replied:
I will update it again once QA has validated it

Progi1984 reviewed (commented):
A small feedback

khouloudbelguith commented:
Hi @eternoendless,

It is ok :heavy_check_mark: 
I make these tests:
1. Install Shop with Arabic language + Saudi Arabia country
2. Add some language via Translation page & pack via Localization page
3. Translate some menus
4. Install Shop with English language + France country
5. Install Shop with French language + France country
6. #9762 => fixed / #9893 (Stock menu) => fixed
7. Add Japan language via Translations page (translate some menus)
8. Import Algeria pack
10. Update all Core Menu 
![image](https://user-images.githubusercontent.com/16067358/114398044-a11bef80-9b9f-11eb-833c-8321e78a408c.png)

PS: https://github.com/PrestaShop/PrestaShop/issues/9816 still exists
https://github.com/PrestaShop/PrestaShop/issues/23156#issuecomment-774157511 still exists

Thank you!

eternoendless commented:
Rebased

eternoendless commented:
Re rebased. Hopefully tests will pass now?

Commit:
Add missing wording and wording_domain for AdminFeatureFlag tab

eternoendless commented:
After rebase, a new Tab was introduced in fixtures, breaking the install.

eternoendless commented:
Thanks everyone! 🎉
